oci/layout/oci_transport.go

```diff
@@ -113,7 +114,7 @@ func NewReference(dir, image string) (types.ImageReference, error) {
 	}
 	// This is necessary to prevent directory paths returned by PolicyConfigurationNamespaces
 	// from being ambiguous with values of PolicyConfigurationIdentity.
-	if strings.Contains(resolved, ":") {
+	if strings.Contains(resolved, ":") && runtime.GOOS != "windows" {
```

Comment:
I am not so sure about this one. Obviously, on Windows, there will be a ':' in the path. Just adding this condition works for my use cases. I am not sure what the comment tries to say in terms of PolicyConfigurationNamespaces and PolicyConfigurationIdentity. I need some guidance here on whether a simple OS check is sufficient here or whether something else needs to happen. Maybe it can at least be a first step solution, since not allowing colon ':' in the path stops this to be working on Windows right in the beginning.

Reply:
The original comment is no longer applicable after 4eb8db31f00f5bcb25fe55341731df33f3067768 (and should be updated, along with `ValidatePolicyConfigurationIdentity`), but this restriction is still somewhat necessary, because the same ambiguity exists in `ParseReference`:

- `/path:busybox` means `dir="/path"`, `image="busybox"`
- `/path` and `/path:` means `dir="/path"`, `image=""`
- `/path:busybox:latest` means `dir="/path"`, `image="busybox:latest`";

Note that `dir="/path:busybox"`, `image="latest"` is impossible to express in the string form consumed by `ParseReference`, because it would be parsed the other way.  Giving up on `ParseReference` for this path and allowing `dir="/path:busybox"` only via `NewReference` would also be problematic because `StringWithinTransport()` must return a valid input for `ParseReference`.

---

I think it would be very reasonable to accept a `:` on Windows (perhaps, to be more strict, only in the second position, but the output of `explicitfilepath.ResolvePathToFullyExplicit` is supposed to be a valid path, and `:` is reserved on Windows, so it should not appear anywhere but after a drive letter.)

The flip side of that, though, is that `ParseReference` should likewise accept paths starting with a drive letter and a colon, and parse that as a part of the `dir` , not as the separator starting `image`.

`ValidatePolicyConfigurationIdentity` should also be updated - to treat the input only as a directory, the way `PolicyConfigurationIdentity` formats it now, and to reject `:` on non-Windows only.

Reply:
I am aware of the problems of ParseReference. In fact, I am not using ParseReference and use NewReference directly. That works fine when using the API programmatically. It is more an issue in the case of Skopeo or other potential tools which want to create a reference from a single string. 

As I mentioned before, I think the use of ':' is unfortunate in this case. I suggested back then to use a different separator like '@' or similar. 

I feel the ParseReference problems is a bigger issue which I'd like to defer to another issue. WDYT?

In moving this along, is a OS specific check ok, together with an adjustment to ValidatePolicyConfigurationIdentity?

Reply:
> I feel the ParseReference problems is a bigger issue which I'd like to defer to another issue. WDYT?

I’d much prefer to keep of `oci_transport.go` consistent. Otherwise there’s no telling which of the alternatives will be updated in the future to make the code consistent, i.e. what use case will break.

An OS-specific check determining the treatment of `:` is OK with me.

Reply:
Another way to put it, if you _really_ don’t want to update `ParseReference` etc. for Windows, I guess I can live with that, but then you’d have to document the code clearly and explicitly enough to keep your work working in the future.

Reply:
@mtrmac any chance you mean _ValidatePolicyConfigurationScope_? There seems to be no _ValidatePolicyConfigurationIdentity_.

If you mean _ValidatePolicyConfigurationScope_, what is the expected input to this function. The same as to _ ParseReference_? Something like <dir>:<image>? 

I am still struggling a bit with these policies and scopes. Not quite sure how everything fits together in this area of the code.

Reply:
@mtrmac another unrelated question. There is one more problem I am seeing which is not addressed in this pull request yet. Could you have a look at https://github.com/containers/image/blob/6e47653e2d514d64af265b69b8cc02542605fccc/copy/copy.go#L290-L290. 

As you know I am exporting a bunch of images from a Docker daemon (running in a VM), store them in an OCI tree on a Windows host and then at some stage re-import into a Docker daemon. This works fine for macOS and Linux. On Windows, I get the error which is referenced above. Effectively: 

    image operating system linux cannot be used on windows

_src.OCIConfig()_ seems to return linux, I would expect Windows though. Any ideas? I have locally commented out the call to _checkImageDestinationForCurrentRuntimeOS_, since I find it questionable either way. Moving forward though, I was wondering how to solve this? If there is no bug in __src.OCIConfig()__ , I was wondering whether it would make sense to add something to SystemContext which allows to skip the check for _ImageDestinationForCurrentRuntimeOS_. WDYT?

Reply:
> `src.OCIConfig()` seems to return `linux`, I would expect `Windows` though

The image was created by a `dockerd` on Linux, which is what is recorded inside the image’s config. It does not matter what OS is accessing the image or where the image was copied, it is designed to run on Linux.

> I was wondering whether it would make sense to add something to `SystemContext` which allows to skip the check for `ImageDestinationForCurrentRuntimeOS`. WDYT?

Yes; I have implemented that already in https://github.com/containers/image/pull/346/commits/8074c3892f3775ec876a63f8e8e2cd5bda9aea5d for https://github.com/containers/image/pull/346 .

Reply:
> Yes; I have implemented that already in 8074c38 for #346 .

LOL :-) Seems to be exactly what I need. Except it does not seem to be merged yet :-( 
I will for now focus on the things we discussed so far and leave the destination check for #346. It would be great to get this merged though. Anything I can do do get #346 merged?

Reply:
> Anything I can do do get #346 merged?

It needs a review by @runcom at least; any extra reviews would be welcome as well.

Reply:
> @mtrmac any chance you mean _ValidatePolicyConfigurationScope_? There seems to be no _ValidatePolicyConfigurationIdentity_.

Yes, sorry.

> If you mean _ValidatePolicyConfigurationScope_, what is the expected input to this function. The same as to _ ParseReference_? Something like `<dir>:<image>`? 
>
> I am still struggling a bit with these policies and scopes. Not quite sure how everything fits together in this area of the code.

1. References: A way for the user to refer to a _single image_ in text, basically a way to express a `types.ImageReference`. There may be multiple ways to refer to a single underlying image, either by definition (`docker://busybox` vs. `docker://busybox:latest`), or by the environment (`dir:pathtodir` vs. `dir:symlinktodir`). Parsed by `Transport.ParseReference`, formatted by `ImageReference.StringWithinTransport` (i.e. the output of this must be consumable by `ParseReference` and be interpreted as the same underlying image).

2. Policy configuration scopes: A way for the user to define a _hierarchical namespace_ of images. The same underlying image must have exactly one, canonical, “identity” (which is why symlinks are resolved in `NewReference`—but note that the identity fundamentally depends on the environment, i.e. the contents of the symlink), and `PolicyConfigurationNamespaces` returns the namespaces which contain that image (i.e. typically the analogue of the set of its parent directories / paths / hostnames etc.). The values are formatted by `PolicyConfigurationIdentity` and `PolicyConfigurationNamespaces`. There is no method to parse the string back into a Go value (strings in config files are compared for equality with the outputs of these two methods), but there is a `ValidatePolicyConfigurationScope` which should allow the output of `PolicyConfiguration{Identity,Scope}` and preferably little or nothing else, to allow detecting errors in the configuration.